src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+                       direction,
+                       sites,
+                       site->getWidth());
+  };
+  create_row("NORTH",
+             vertical_site,
+             x_sites,
+             {nw->getBBox().xMax(),
+              outer_io.yMax() - static_cast<int>(vertical_site->getHeight())},
+             odb::dbOrientType::MX,
+             odb::dbRowDir::HORIZONTAL);
+  create_row("EAST",
+             horizontal_site,
+             y_sites,
+             {outer_io.xMax() - static_cast<int>(horizontal_site->getHeight()),
+              se->getBBox().yMax()},
+             odb::dbOrientType::R90,
+             odb::dbRowDir::VERTICAL);
+  create_row("SOUTH",
+             vertical_site,
+             x_sites,
+             {sw->getBBox().xMax(), outer_io.yMin()},
+             odb::dbOrientType::R0,
+             odb::dbRowDir::HORIZONTAL);
+  create_row("WEST",
+             horizontal_site,
+             y_sites,
+             {outer_io.xMin(), sw->getBBox().yMax()},
+             odb::dbOrientType::MXR90,
+             odb::dbRowDir::VERTICAL);
+}
+
+void ICeWall::removeIORows()
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  for (auto* row : getRows()) {
+    odb::dbRow::destroy(row);
+  }
+}
+
+void ICeWall::placeCorner(odb::dbMaster* master, int ring_index)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (master == nullptr) {
+    return;
+  }
+
+  for (const char* row_name : {"IO_CORNER_NORTH_WEST",
+                               "IO_CORNER_NORTH_EAST",
+                               "IO_CORNER_SOUTH_WEST",
+                               "IO_CORNER_SOUTH_EAST"}) {
+    odb::dbRow* row;
+    if (ring_index >= 0) {
+      row = findRow(fmt::format("{}_{}", row_name, ring_index));
+    } else {
+      row = findRow(row_name);
+    }
+    if (row == nullptr) {
+      continue;
+    }
+
+    const std::string corner_name = fmt::format("{}_INST", row->getName());
+    auto* inst = odb::dbInst::create(block, master, corner_name.c_str());
+
+    const odb::Rect row_bbox = row->getBBox();
+
+    inst->setOrient(row->getOrient());
+    inst->setLocation(row_bbox.xMin(), row_bbox.yMin());
+    inst->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+  }
+}
+
+void ICeWall::placePad(odb::dbMaster* master,
+                       const std::string& name,
+                       odb::dbRow* row,
+                       int location,
+                       bool mirror)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  odb::dbInst* inst = block->findInst(name.c_str());
+  if (inst == nullptr) {
+    if (master == nullptr) {
+      return;
+    }
+    inst = odb::dbInst::create(block, master, name.c_str());
+  }
+
+  if (row == nullptr) {
+    return;
+  }
+
+  odb::dbTransform orient(odb::dbOrientType::R0);
+  if (mirror) {
+    const odb::dbTransform mirror_transform(odb::dbOrientType::MY);
```

Comment:
why does mirror == MY?  Do we not mirror with MX on the other sides?

Reply:
For the other sides, the MY is concatenated with the row rotation, so a R90 row will end up with a MX rotation.